File: src/action/DeviceRequest.php
<?php


namespace Ething\Action;


use \Ething\Device\Device;
use \Ething\StreamBuffer;
use \Ething\Helpers;

class DeviceRequest extends Action {
	
	static public function validate(array &$attributes, array $context){
		
		$attributes = array_merge(
			array( 
				'device' => null, // device id
				'operation' => null, // operation id
				'parameters' => null,
				'output' => null
			),
			$attributes
		);
		
		$device = null;
		
		foreach(array_keys($attributes) as $key){
			
			switch($key){
				
				case 'device':
					
					if(!is_string($attributes[$key]))
						throw new \Exception("field '{$key}' must be a string.");
					
					$device = $context['ething']->get($attributes[$key]);
					if(!$device || !($device instanceof Device))
						throw new \Exception("the resource with id '{$attributes[$key]}' does not exist or is not a Device.");
					
					break;
				
				case 'operation':
					
					if(empty($attributes[$key]) ||  !is_string($attributes[$key]))
						throw new \Exception("field '{$key}' must be a non empty string.");
					
					break;
					
				case 'parameters':
					
					if(is_object($attributes[$key]))
						$attributes[$key] = (array)$attributes[$key];
					
					if(!is_array($attributes[$key]) && ! is_null($attributes[$key]))
						throw new \Exception("field '{$key}' is invalid.");
					
					break;
				
				case 'output':
					
					$value = &$attributes[$key];
					
					if(is_object($value))
						$value = (array)$value;
					
					if(is_array($value) && is_string($value['type'])){
						
						$o = array(
							'type' => $value['type'],
							'expireAfter' => null
						);
						
						switch($value['type']){
							case 'file':
							case 'table':
								
								if(empty($value['name']) || !is_string($value['name']))
									throw new \Exception("field '{$key}.name' must be a non empty string.");
								
								$o['name'] = $value['name'];
								
								if(isset($value['expireAfter'])){
									if(!(is_int($value['expireAfter']) && $value['expireAfter']>=0))
										throw new \Exception("field '{$key}.expireAfter' must be a positive integer.");
									$o['expireAfter'] = $value['expireAfter'];
								}
								
								
								break;
							
							default:
								throw new \Exception("field '{$key}.type' : unknown.");
								
						}
						
						$value = $o;
						
					}
					else if(!is_null($value))
						throw new \Exception("field '{$key}' is invalid.");
					
					break;
					
				default:
					throw new \Exception("field '{$key}' unknown.");
			}
			
		}
		
		
		$operation = $device->operation($attributes['operation']);
		if(!$operation)
			throw new \Exception("operation '{$attributes['operation']}' does not exist.");
		
		
		return true; 
	}
	
	
	protected function call(\Ething\Event\Signal $signal){
		
		$device = $this->ething()->get( $this->device );
		if(!$device)
			throw new \Ething\InvalidRuleException("The device #{$this->device} does not exist any more");
		
		$operation = $device->operation($this->operation);
		if(!$operation)
			throw new \Exception("The device {$this->device} has no operation '{$operation}'");
		
		$buffer = new StreamBuffer();
		
		$res = $operation->call($buffer, $this->parameters, array(
			'timeout' => 5000 // 5 secondes
		));
		
		if(!$res){
			throw new \Exception($buffer->errData());
		}
		
		if(!is_null($this->output) && $buffer->length()){
			
			switch($this->output['type']){
				
				case 'file':
					
					$self = $this;
					$filename = Helpers::percent_replace($this->output['name'], function($tag) use ($signal, $self) {
						switch($tag){
							case '%%':
								return '%';
							case '%D':
								return date("Y-m-d H:i:s", $signal->getTimestamp());
							case '%R':
								if(isset($signal->resource)){
									if($r = $self->ething()->get($signal->resource))
										return $r->name();
								}
								break;
							case '%I':
								if(isset($signal->resource)){
									return $r->resource;
								}
								break;
							case '%r':
								return $self->rule()->name;
								break;
							case '%s':
								return $signal->getName();
								break;
							case '%c':
								return (string)($self->rule()->executedCount);
								break;
						}
					});
					
					// does this file already exist !
					$file = $this->ething()->findOne(array(
						'type' => 'File',
						'name' => $filename,
						'createdBy.id' => $device->id()
					));
					if(!$file){
						// no, create it !
						$opt = array( 'name' => $filename );
						if(!empty($this->output['expireAfter']))
							$opt['expireAfter'] = $this->output['expireAfter'];
						$file = $this->ething()->create('File', $opt, $device);
					}
					if($file){
						$file->write( $buffer->data() ); // put the content in it !
					}
					
					break;
				
				case 'table':
						
					// get the data from the buffer
					$data = $buffer->decode($err);
					if(!$err){
						
						// does the table already exist ?
						$table = $this->ething()->findOne(array(
							'name' => $this->output['name'],
							'createdBy.id' => $device->id()
						));
						
						if(!$table){
							// no, so create it !
							$table = $this->ething()->create('Table', array( 'name' => $this->output['name'] ), $device);
						}
						if($table){
						
							if(is_int($data) || is_float($data) || is_string($data) || is_bool($data) || is_null($data)) // transform scalar value into array
								$data = array(
									'value' => $data
								);
							
							$table->insert($data);
						}
					}
					
					break;
				
			}
		}
		
	}
	
}

File: ething/rule/action/DeviceRequest.py

from .. import Action
from .. import InvalidRuleException


class DeviceRequest(Action):
	
	@staticmethod
	def validate (attributes, context):
		
		attributes.setdefault('device', None)
		attributes.setdefault('operation', None)
		attributes.setdefault('parameters', None)
		attributes.setdefault('output', None)
		
		dev = None
		
		for key in attributes:
			
			value = attributes[key]
			
			if key == 'device':
				
				if not isinstance(value, basestring):
					raise Exception("must be a device id.")
				
				dev = context['ething'].get(value)
				
				if not dev or not dev.isTypeof('Device'):
					raise Exception("the resource with id '%s' does not exist or is not a Device." % value)
				
				
			elif key == 'operation':
				
				if not( isinstance(value, basestring) and len(value)>0 ):
					raise Exception("must be a non empty string.")
			
			elif key == 'parameters':
				if not isinstance(value, dict) and value is not None:
					raise Exception("%s: invalid" % key)
			
			elif key == 'output':
				
				if not isinstance(value, dict) or 'type' not in value or not isinstance(value['type'], basestring) or 'name' not in value or not isinstance(value['name'], basestring):
					raise Exception("%s: invalid" % key)
				
				o = {
					'type': value['type'],
					'expireAfter': None,
					'name': value['name']
				}
				
				if value['type'] == 'File' or value['type'] == 'Table':
					
					if 'expireAfter' in value :
						if not isinstance(value['expireAfter'], int) or value['expireAfter']<0:
							raise Exception("%s: invalid" % key)
						o['expireAfter'] = value['expireAfter']
					
				else:
					raise Exception("output type can only File or Table.")
				
				attributes[key] = o
				
			else:
				raise Exception("%s: invalid" % key)
		
		if attributes['operation'] not in dev.interface.method_names:
			raise Exception("operation '%s' does not exist." % attributes['operation'])
		
		
		return True
	
	
	
	def call(self, signal):
		
		device = self.ething.get(self['device'])
		if not device:
			raise InvalidRuleException("The device %s does not exist any more" % str(device))
		
		kwargs = {}
		if self['parameters']:
			kwargs.update(self['parameters'])
		
		result = device.interface.call(self['operation'], **kwargs)
		
		output = self['output']
		if output:
			
			filename = output['name']
			
			if output['type'] == 'File':
				
				#does this file already exist !
				file = self.ething.findOne({
					'type' : 'File',
					'name' : filename,
					'createdBy' : device.id
				})
				
				if not file:
					# no, create it !
					opt = { 'name' : filename }
					if output['expireAfter'] is not None :
						opt['expireAfter'] = output['expireAfter']
					file = self.ething.create('File', opt, device)
				
				if file:
					file.write( str(result) )
				
			elif output['type'] == 'Table':
				
				#does this file already exist !
				table = self.ething.findOne({
					'type' : 'Table',
					'name' : filename,
					'createdBy' : device.id
				})
				
				if not table:
					# no, create it !
					opt = { 'name' : filename }
					if output['expireAfter'] is not None :
						opt['expireAfter'] = output['expireAfter']
					table = self.ething.create('Table', opt, device)
				
				if table:
					
					if isinstance(result, int) or isinstance(result, float) or isinstance(result, basestring) or isinstance(result, bool) or result is None:
						table.insert({
							'value': result
						})
					elif isinstance(result, dict):
						table.insert(result)
					else:
						raise Exception('unable to put the data in the output table')
			
		
	
	
	